### PR TITLE
[DF] Improve reading collections with RNTuple DS

### PR DESCRIPTION
**Pull request by jblomer**
Adds the following columns to an RDF created from an RNTuple:

Inner collections
----------------------
The inner fields of collection of a record is projected to collections of the inner fields.  E.g. a for a field `jet` of type vector of

    struct Jet {
      float pt;
      float eta;
    };

the RDF exposes `std::vector<float> jet.pt` and `std::vector<float> jet.eta` in addition to `std::vector<Jet> jet`

Cardinality columns
--------------------------
For every collection `x`, an additional cardinality column `#x` is exposed that contains the number of elements of the collection.  Note that for inner collections, the cardinality columns is itself a collection. For instance, for a column `std::vector<std::vector<float>>` the cardinality column is of type `std::vector<std::uint32_t>`.

### CODE REVIEW

**Comment by jblomer** [comment]
@phsft-bot build

**Comment by jblomer** [comment]
@phsft-bot build

**Review by eguiraud — CHANGES_REQUESTED** [review]
Thank you @jblomer , this is major! And sorry for the delay in the review.

Two main comments/questions:

1. does the `#column` notation work with just-in-time compilation? I am not sure how our tokenizer treats it. I don't think we have tests with RNTupleDS + jitting, e.g. `df.Define("xlength", "#x").Histo1D("#x")`, can you please add some?

2. is the plan to use RVecs instead of std::vectors in the future? if yes, can you please leave a comment in the relevant places?

P.S.
I have a hard time understanding some parts of `RNTupleDS::AddProjection`. Can you point me to docs/tutorials that explain usage of `Connect` vs `ConnectRecursively` and `GetFieldRange`?

File: tree/dataframe/src/RNTupleDS.cxx [hunk 1]
```diff
@@ -33,31 +33,83 @@
...
+class RNTupleColumnReader : public ROOT::Detail::RDF::RColumnReaderBase {
+protected:
    using RFieldBase = ROOT::Experimental::Detail::RFieldBase;
    using RFieldValue = ROOT::Experimental::Detail::RFieldValue;
    using RPageSource = ROOT::Experimental::Detail::RPageSource;
 
-   std::unique_ptr<RFieldBase> fField;
-   RFieldValue fValue;
+   std::unique_ptr<RFieldBase> fField; ///< The field backing the RDF column
+   std::vector<DescriptorId_t> fSkeinIDs; ///< For inner collections, the field IDs of the outer collections
```
Comment (eguiraud) [inline tree/dataframe/src/RNTupleDS.cxx:97]:
```suggestion
   /// For inner collections, the field IDs of the outer collections, from outermost to innermost. Empty otherwise.
   std::vector<DescriptorId_t> fSkeinIDs;
```

**Comment by jblomer** [comment]
>     1. does the `#column` notation work with just-in-time compilation? I am not sure how our tokenizer treats it. I don't think we have tests with RNTupleDS + jitting, e.g. `df.Define("xlength", "#x").Histo1D("#x")`, can you please add some?

Pity!  I get an error with
```
auto lambda0 = [](){return #jets
```

File: tree/dataframe/src/RNTupleDS.cxx [hunk 2]
```diff
@@ -70,13 +68,23 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
...
+
+
+void RNTupleDS::AddRecord(const RNTupleDescriptor &desc, DescriptorId_t parentId)
 {
    for (const auto& f : desc.GetFieldRange(parentId)) {
       fColumnNames.emplace_back(desc.GetQualifiedFieldName(f.GetId()));
       fColumnTypes.emplace_back(f.GetTypeName());
+
+      auto field = Detail::RFieldBase::Create(f.GetFieldName(), f.GetTypeName());
+      auto columnReader = std::make_unique<Detail::RNTupleColumnReader>(std::move(field), f.GetId());
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx]: As discussed, the work of figuring out the RDF columns and its types has to be done at data source construction time.  We invest a little bit more work and at the same time also create the backing fields for each RDF column and store it as a column reader prototype.  We can then clone the prototype once per slot and used RDF column.

File: tree/dataframe/inc/ROOT/RNTupleDS.hxx [hunk 3]
```diff
@@ -54,8 +54,10 @@ class RNTupleDS final : public ROOT::RDF::RDataSource {
    unsigned fNSlots = 0;
    bool fHasSeenAllRanges = false;
 
-   void AddCollection(const RNTupleDescriptor &desc, DescriptorId_t collectionId);
-   void AddRecord(const RNTupleDescriptor &desc, DescriptorId_t recordId);
+   void AddProjection(const RNTupleDescriptor &desc,
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/inc/ROOT/RNTupleDS.hxx]: Renamed to `AddField`

File: tree/dataframe/src/RNTupleDS.cxx [hunk 4]
```diff
@@ -68,27 +120,122 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
...
+
+   std::unique_ptr<RNTupleColumnReader> Clone() const final {
+      return std::make_unique<RNTupleCardinalityColumnReader>(fField->Clone(fField->GetName()), fSkeinIDs);
+   }
+
+   void Connect(RPageSource &source) final {
+      auto f = fField.get();
+      for (unsigned i = 0; i < fSkeinIDs.size() - 1; ++i) {
+         Detail::RFieldFuse::Connect(fSkeinIDs[i], source, *f);
+         f = f->GetSubFields()[0];
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx:142]: Comment added

File: tree/dataframe/src/RNTupleDS.cxx [hunk 5]
```diff
@@ -68,27 +120,122 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
...
+      }
+      auto cardinalityField = dynamic_cast<RRDFCardinalityField *>(f);
+      Detail::RFieldFuse::Connect(fSkeinIDs.back(), source, cardinalityField->fOffsetField);
+   }
+};
+
+/// Provides a particular field of the RNTuple as RDF column
+class RNTupleProjectionColumnReader : public RNTupleColumnReader {
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx]: Renamed to `RNTupleNormalColumnReader`

File: tree/dataframe/src/RNTupleDS.cxx [hunk 6]
```diff
@@ -68,27 +120,122 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
...
+};
+
 } // namespace Detail
 
-void RNTupleDS::AddFields(const RNTupleDescriptor &desc, DescriptorId_t parentId)
+
+RNTupleDS::~RNTupleDS() = default;
+
+
+void RNTupleDS::AddProjection(
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx]: I added a concrete example to `AddField()`.  Let me know if that makes it more clear.

File: tree/dataframe/src/RNTupleDS.cxx [hunk 7]
```diff
@@ -33,7 +33,47 @@
...
+   }
+   size_t GetValueSize() const final { return sizeof(ClusterSize_t); }
+
+   void ReadGlobalImpl(ROOT::Experimental::NTupleSize_t globalIndex,
+                       ROOT::Experimental::Detail::RFieldValue *value) final
+   {
+      RClusterIndex collectionStart;
+      fIndexField.GetCollectionInfo(globalIndex, &collectionStart, value->Get<ClusterSize_t>());
+   }
+};
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx:90]: Comment added.

**Comment by eguiraud** [comment]
@phsft-bot build please

File: tree/dataframe/inc/ROOT/RNTupleDS.hxx [hunk 8]
```diff
@@ -35,26 +35,41 @@ class RNTupleDescriptor;
...
 
-   void AddFields(const RNTupleDescriptor &desc, DescriptorId_t parentId);
+   /// Provides the RDF column "colName" given the field identified by fieldID. For records and collections,
+   /// AddField recurses into the sub fields. The skeinIDs is the list of field IDs of the outer collections
+   /// of fieldId. For instance, if fieldId refers to an `std::vector<Jet>`, with
+   /// struct Jet {
+   ///    float pt;
+   ///    float eta;
+   /// };
+   /// AddProject will recurse into Jet.pt and Jet.eta and provide the two inner fields as std::vector<float> each.
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RNTupleDS.hxx:64]:
```suggestion
   /// AddField will recurse into Jet.pt and Jet.eta and provide the two inner fields as std::vector<float> each.
```

File: tree/dataframe/test/datasource_ntuple.cxx [hunk 9]
```diff
@@ -47,14 +47,26 @@ TEST_F(RNTupleDSTest, ColTypeNames)
...
+   EXPECT_STREQ("ROOT::Experimental::ClusterSize_t::ValueType", tds.GetTypeName("#jets").c_str());
+}
+
+
+TEST_F(RNTupleDSTest, CardinalityColumn)
+{
+   auto df = ROOT::Experimental::MakeNTupleDataFrame(fNtplName, fFileName);
+
+   // Check that the special column #<collection> works with jitting
+   auto max_njets = df.Define("njets", "@jets.size").Max("njets");
```
Comment (eguiraud) [inline tree/dataframe/test/datasource_ntuple.cxx:68]:
```suggestion
   auto max_njets = df.Define("njets", "#jets").Max("njets");
```

Comment (jblomer) [reply tree/dataframe/test/datasource_ntuple.cxx:68]: With respect to the tokenizer, that doesn't help us unfortunately. The `@` symbol is equally invalid syntax.

Comment (eguiraud) [reply tree/dataframe/test/datasource_ntuple.cxx:68]: yes yes, just to remind ourselves to change it back to `#`

**Comment by jblomer** [comment]
#7341 should make this PR a little easier

**Comment by jblomer** [comment]
Fix-up taking into account #7354 and #7341

File: tree/dataframe/inc/ROOT/RNTupleDS.hxx [hunk 10]
```diff
@@ -35,26 +35,41 @@ class RNTupleDescriptor;
...
 
-   void AddFields(const RNTupleDescriptor &desc, DescriptorId_t parentId);
+   /// Provides the RDF column "colName" given the field identified by fieldID. For records and collections,
+   /// AddField recurses into the sub fields. The skeinIDs is the list of field IDs of the outer collections
+   /// of fieldId. For instance, if fieldId refers to an `std::vector<Jet>`, with
+   /// struct Jet {
+   ///    float pt;
+   ///    float eta;
+   /// };
+   /// AddField will recurse into Jet.pt and Jet.eta and provide the two inner fields as std::vector<float> each.
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/inc/ROOT/RNTupleDS.hxx:64]: Yep

**Comment by eguiraud** [comment]
Other general comments:

1. I know we discussed this already, but I can't remember the full discussion, sorry: why is it better to create column reader prototypes upfront rather than doing so on-demand? in particular, what is the expensive operation that we perform less times?

2. Given that the last commits change the logic of the previous ones drastically, this PR might be a good candidate for a squash on merge or at least some of the commits are good candidates to be squashed together.

File: tree/dataframe/src/RNTupleDS.cxx [hunk 11]
```diff
@@ -69,27 +133,120 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
...
+      valueField->SetOnDiskId(*i);
+      // Skip the inner-most collection level to construct the cardinality column
+      if (i != skeinIDs.rbegin()) {
+         cardinalityField = std::make_unique<ROOT::Experimental::RVectorField>("", std::move(cardinalityField));
+         cardinalityField->SetOnDiskId(*i);
+      }
+   }
+
+   if (cardinalityField) {
+      fColumnNames.emplace_back(std::string("__rdf_sizeof_") + std::string(colName));
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx:228]: Are you sure this works? I often see issue adding an `std::string` to a C string literal because the compiler thinks I want to add an `std::string` to a pointer

[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/src/RNTupleDS.cxx:228]: yep `operator+` between std::string and string literal is "commutative"

File: tree/dataframe/src/RNTupleDS.cxx [hunk 12]
```diff
@@ -69,27 +133,120 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
...
+   // "event"                             [Event]
+   // "event.id"                          [int]
+   // "event.tracks"                      [std::vector<Track>]
+   // "__rdf_sizeof_event.tracks"         [unsigned int]
+   // "event.tracks.hits"                 [std::vector<std::vector<Hit>>]
+   // "__rdf_sizeof_event.tracks.hits"    [std::vector<unsigned int>]
+   // "event.tracks.hits.x"               [std::vector<std::vector<float>>]
+   // "__rdf_sizeof_event.tracks.hits.x"  [std::vector<unsigned int>]
+   // "event.tracks.hits.y"               [std::vector<std::vector<float>>]
+   // "__rdf_sizeof_event.tracks.hits.y"  [std::vector<unsigned int>]
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx:176]: The problem is: which size? The sum of hit sizes over all tracks?

File: tree/dataframe/src/RNTupleDS.cxx [hunk 13]
```diff
@@ -33,31 +33,95 @@
...
-class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
+
+/// An artificial field that transforms an RNTuple column that contains the offset of collections into
+/// collection sizes. It is used to provide the "number of" RDF columns for collections, e.g.
+/// `__rdf_sizeof_jets` for a collection named `jets`.
+///
+/// This field owns the collection offset field but instead of exposing the collection offsets it exposes
+/// the collection sizes (offset(N+1) - offset(N)).  For the time being, we offer this functionality only in RDataFrame.
+/// TODO(jblomer): consider providing a general set of useful virtual fields as part of RNTuple.
+class RRDFCardinalityField : public ROOT::Experimental::Detail::RFieldBase {
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx:44]: Done

File: tree/dataframe/src/RNTupleDS.cxx [hunk 14]
```diff
@@ -33,31 +33,95 @@
...
+   }
+
+public:
+   static std::string TypeName() { return "ROOT::Experimental::ClusterSize_t::ValueType"; }
+   RRDFCardinalityField()
+     : Detail::RFieldBase("", TypeName(), ENTupleStructure::kLeaf, false /* isSimple */)
+   {
+   }
+   RRDFCardinalityField(RRDFCardinalityField&& other) = default;
+   RRDFCardinalityField& operator =(RRDFCardinalityField&& other) = default;
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx]: Done by a `clang-format` run

**Comment by jblomer** [comment]
Ad 1) I find it simpler to code to build all the fields upfront. We'll save a bit of memory doing it on demand, but it will possibly take a little longer because we need to traverse the schema tree more often (for every leaf, we have to figure out again what are the parents).

Ad 2) I'm fine with a squash on merge.  I'd prefer not to squash individual commits, it's a long history and a partial rewrite will get it as wrong as the full history.  Let me know if you like the squash commit over the rebase.

File: tree/dataframe/src/RNTupleDS.cxx (hunk 12 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/src/RNTupleDS.cxx:171]: My mental model is "`#var` is a performant shorthand for `var.size()`", so `#event.tracks.hits` would actually be equivalent to `#event.tracks` (because the `vector<Track>` and the `vector<vector<float>>` return the same number when you call `.size()` on them).

Your version probably makes more sense :smile: but I don't know how to explain it in plain english to users.

**Comment by eguiraud** [comment]
In my opinion squashing might make more sense for this PR, but up to you, it's not a merge-blocking thing as long as each commit compiles (which I believe is the case).

**Comment by jblomer** [comment]
Oh, actually I remember one more reason why we build the fields beforehand: we anyway have to figure out beforehand which column _names_ exist. Figuring that out is almost the same than going all the way and creating the fields.

Good point on the whether or not it compiles! There are commits that don't. So squash it is.

**Review by eguiraud — APPROVED** [review]
I think this is good to go

File: tree/dataframe/src/RNTupleDS.cxx (hunk 12 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/src/RNTupleDS.cxx:171]: will have to be documented very well, but it makes sense as it is

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by jblomer** [comment]
Almost there... I'll take a look into the tutorial failure.
